risc0/zkvm/src/host/client/prove/bonsai.rs

```diff
@@ -85,7 +85,18 @@ impl Prover for BonsaiProver {
 
                 let receipt_buf = client.download(&receipt_url)?;
                 let receipt: Receipt = bincode::deserialize(&receipt_buf)?;
-                receipt.verify_with_context(ctx, image_id)?;
+
+                if opts.prove_guest_errors {
+                    receipt.verify_integrity_with_context(ctx)?;
+                    ensure!(
+                        receipt.get_metadata()?.pre.digest() == image_id,
+                        "received unexpected image ID: expected {}, found {}",
+                        hex::encode(&image_id),
+                        hex::encode(&receipt.get_metadata()?.pre.digest())
+                    );
+                } else {
+                    receipt.verify_with_context(ctx, image_id)?;
+                }
```

Comment:
I wonder why this doesn't belong in `verify_with_context` on the receipt, rather than for each caller to do?

Reply:
I wouldn't expect most callers to have a if statement like this. It's in this location because the Bonsai client is implementing `Prover` and this is where it is checking the Bonsai's work. So the conditions that need to be enforced here depend on whether the `Prover` interface user is expecting a successful execution or doesn't care.